src/Owin.Security.Providers.ConstantContact/ConstantContactAuthenticationHandler.cs

```diff
@@ -0,0 +1,225 @@
+                properties.Dictionary.Add(new KeyValuePair<string, string>(StateCookie, protectedRequestToken));
+
+                values = query.GetValues("username");
+                if (values != null && values.Count == 1)
+                {
+                    user = values[0];
+                }
+
+                // OAuth2 10.12 CSRF
+                if (!ValidateCorrelationId(properties, _logger))
+                {
+                    return new AuthenticationTicket(null, properties);
+                }
+
+                var requestPrefix = Request.Scheme + "://" + Request.Host;
+                var redirectUri = requestPrefix + Request.PathBase + Options.CallbackPath;
+
+                // Build up the body for the token request
+                var body = new List<KeyValuePair<string, string>>
+                {
+                    new KeyValuePair<string, string>("code", code),
+                    new KeyValuePair<string, string>("grant_type", "authorization_code"),
+                    new KeyValuePair<string, string>("client_id", Options.ClientId),
+                    new KeyValuePair<string, string>("client_secret", Options.ClientSecret),
+                    new KeyValuePair<string, string>("redirect_uri", redirectUri)
+                };
+
+                // Request the token
+                var requestMessage = new HttpRequestMessage(HttpMethod.Post, Options.Endpoints.TokenEndpoint);
+                requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                requestMessage.Content = new FormUrlEncodedContent(body);
+                var tokenResponse = await _httpClient.SendAsync(requestMessage);
+                tokenResponse.EnsureSuccessStatusCode();
+                var text = await tokenResponse.Content.ReadAsStringAsync();
+
+                // Deserializes the token response
+                dynamic response = JsonConvert.DeserializeObject<dynamic>(text);
+                var accessToken = (string)response.access_token;
+                var expires_in = (string)response.expires_in;
+
+                properties.IsPersistent = true;
+                properties.IssuedUtc = DateTime.Now;
+                properties.RedirectUri = redirectUri;
+
+                // Get the user info
+
+                var context = new ConstantContactAuthenticatedContext(Context, accessToken, user, expires_in)
+                {
+                    Identity = new ClaimsIdentity(
+                        Options.AuthenticationType,
+                        ClaimsIdentity.DefaultNameClaimType,
+                        ClaimsIdentity.DefaultRoleClaimType)
+                };
+                if (!string.IsNullOrEmpty(context.AccessToken))
+                {
+                    context.Identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, context.AccessToken, XmlSchemaString, Options.AuthenticationType));
```

Comment:
I don't believe any of the providers in this repository place their access tokens right in the users claims. If people wish to get access to these tokens they can pass in their own AuthenticationProvider with its own OnAuthenticated lambda 

http://stackoverflow.com/questions/18942196/how-to-access-facebook-private-information-by-using-asp-net-identity-owin/19549581#19549581

https://github.com/TerribleDev/OwinOAuthProviders/issues/176#issuecomment-241876404